### ISSUE TITLE
Error when edit category in case of more then one language is installed

### ISSUE THREAD

**Issue by nmpetkov** [zikula/core#3124]
| Q | A |
| --- | --- |
| Zikula Version | 1.4.3 |
| PHP Version | 5.6.24 |
### Expected behaviour

No error when edit category in case of more then one language is installed
### Actual behaviour

Obtain an error when trying to edit a category in case of more then one language is installed.
### Steps to reproduce

Install Zikula 1.4.3, then install one more language.
Go to Administration, Categories, and expand the list. Try to edit a category.
Error:
`Key "bg" for array with keys "en" does not exist in system/CategoriesModule/Resources/views/Admin/edit.html.twig at line 68`
### Comment:

One possible solution is to check if display_name is set for each language in function getCategoryByID in file lib/util/CategoryUtil.php. Find line
`$cat = $category->toArray();`
and insert after:

```
        // set display_name by languages if not set
        $languages = ZLanguage::getInstalledLanguages();
        foreach ($languages as $lang) {
            if (!isset($cat['display_name'][$lang])) {
                $cat['display_name'][$lang] = isset($cat['display_name']['en']) ? $cat['display_name']['en'] : '';
            }
            if (!isset($cat['display_desc'][$lang])) {
                $cat['display_desc'][$lang] = isset($cat['display_desc']['en']) ? $cat['display_desc']['en'] : '';
            }
        }
```

Also in function getCategories.

**Comment by nmpetkov** [comment]
Some more info about this issue: The error occurs if the additional language is installed AFTER Zikula installation. If the language files are added to Zikula package BEFORE to upload files and to install Zikula, then no error (I tested this case).
Anyway it is normal practice to install additional languages AFTER installing Zikula.

**Comment by craigh** [comment]
@Kaik would you know anything about this? Is there a scan on install for languages?

**Comment by nmpetkov** [comment]
Sorry, but please neglect my previous comment (about order of installed languages).

Now after more testing and new install, I found the error occurs only if set
`debug: true`
in app/config/custom_parameters.yml file.
(I'm switching to debug mode in case I have an error and want to debug).

**Comment by Kaik** [comment]
Hi, each time ZLanguage::getInstalledLanguages() is called it is scanning for languages here https://github.com/zikula/core/blob/1.4/src/lib/i18n/ZLanguage.php#L646
@craigh Yes there is scan during installation it could populate languages dropdown if there was no scan.
@nmpetkov I think it is an issue this should work without error in debug mode as well.

Looks like the problem is in template loop here and in similar loops below as well https://github.com/zikula/core/blob/1.4/src/system/CategoriesModule/Resources/views/Admin/edit.html.twig#L68

It is because `` languages `` array have all language's that are installed (including those installed recently) while `` category.display_name `` array contains only "old indexes" and edit function is the only place where new indexes can be added.

It can be solved adding a check if array key exist in those set functions in template.

**Comment by Guite** [comment]
@Kaik first I assumed that it would be required to add a dedicated check like this:

```
{% set displayName = category.display_name[language] is defined ? category.display_name[language] : '' %}
...
{% set displayDesc = category.display_desc[language] is defined ? category.display_desc[language] : '' %}
```

But then my second thought was that a default modifier should do the trick less verbosely.

Could you please test the proposed change in #3135?
If this works fine. If not, could you please also test the approach I've pasted above?
Thanks.

**Comment by Guite** [comment]
@nmpetkov could you please test the two proposals from my previous comment?

**Comment by Guite** [comment]
Merged #3135

**Comment by nmpetkov** [comment]
I tested again with Zikula 1.4.4.build100 fresh install with ```debug: true``` in app/config/custom_parameters.yml file, and one additional language installed.
The result is:
1. Now edit a category with single click on the name works. This is case when the edit occurs in full page.
2. **NOT WORKING**: Try to edit in AJAX mode (when right click on the name of a category). You obtain ```500 Internal Server Error``` in a popup window.
3. WORKING: Try to edit in AJAX mode after this category is previously edited in full page (that means in database language strings are written already).

Then I changed file ```lib/util/CategoryUtil.php``` in the manner I proposed in first post in this issue (check if display_name is set for each language, and set). After this AJAX edit works fine even this category is not edited before.
